src/coreclr/jit/gentree.cpp

```diff
@@ -11702,6 +11702,66 @@ GenTree* Compiler::gtFoldExprCall(GenTreeCall* call)
     return call;
 }
 
+#ifdef FEATURE_HW_INTRINSICS
+
+//------------------------------------------------------------------------
+// gtFoldHWIntrinsic: see if a HWIntrinsic is foldable
+//
+// Arguments:
+//    node - HWIntrinsic node to examine
+//
+// Returns:
+//    The original node if no folding happened.
+//    An alternative tree if folding happens.
```

Comment:
Nit: the "return a different (by pointer value) tree pattern to detect "did you optimize" is not great, since it makes bashing more cumbersome. I think it would be better to return `nullptr` on "failure" instead.

Reply:
Interesting, I don't know much about bashing. The pattern that I'm following here is from `gtFoldExprCall` as it does the same thing.